### DISCUSSION TITLE
Open popup from view model

### DISCUSSION THREAD

**Discussion by imsam67** [CommunityToolkit/Maui#566] [Q&A] [answered]
Is there a way to open popup from a viewmodel? The documentation indicates that it can only be opened from a ContentPage but I was wondering if there’s a workaround.

**Comment by VladislavAntonyuk** [comment]
Pop up is UI element, but you can still call it from view model:
Application.Current.MainPage.ShowPopup(popup)

**Comment by kphillpotts** [comment, marked as answer]
I think it depends on how much of an MVVM-Purist™ you are. 😉 I'm a pragmatist, so personally I'd be more than happy to go with the approach suggested by @VladislavAntonyuk.

But if you are a purist and want to decouple UI elements from your ViewModels then another option would be to create a IPopupService which you inject into your ViewModels. Something like this:

**create interface**
``` csharp
public interface IPopupService
{
	void ShowPopup(Popup popup);
}
```
**create implementation**
``` csharp
public class PopupService : IPopupService
{
	public void ShowPopup(Popup popup)
	{
		Page page = Application.Current?.MainPage ?? throw new NullReferenceException();
		page.ShowPopup(popup);
	}
}
```
**don't forget to register it**
``` csharp
services.AddTransient<IPopupService, PopupService> ();
```
**inject service into your viewmodel and use it**
``` csharp
	public PopupAnchorViewModel(IPopupService popupService)
	{
		this.popupService = popupService;
	}

        // and then somewhere (probably in a command)...
        popupService.ShowPopup(popup);	
```

But that's a lot of code, and in reality, you probably aren't really achieving much. 

So yeah, for my money you could go with what @VladislavAntonyuk said, which is pretty much how it's done in the [Sample App ViewModels](https://github.com/CommunityToolkit/Maui/blob/0cd83f96cdb9b9776ac3276e0ed6e141374e6161/samples/CommunityToolkit.Maui.Sample/ViewModels/Views/Popup/PopupAnchorViewModel.cs#L9). I just thought it would be interesting to discuss an alternate approach.

**Reply by VladislavAntonyuk** [reply]
In my production app I use the implementation which you suggested, because I can always replace the dialog service implementation without touching the whole app where direct Popup service is used. But anyway because we depends on Page which is UI element, you will probably implement you DialogService in MAUI app

**Reply by imsam67** [reply]
Thank you all for your help! Works nicely and I really appreciate the detailed example of creating a service. I really like that approach!

**Reply by bijington** [reply]
I guess if we are talking about Purists someone might argue that the `Popup` is view based and we should abstract that away 🤣

**Reply by kphillpotts** [reply]
That's why I'm not an MVVM Purist™

**Comment by CliffAgius** [comment]
Closed as answered
